### ISSUE TITLE
Low-latency streaming

### ISSUE THREAD

**Issue by dantmnf** [rustdesk/rustdesk#79]
The current screen-to-screen latency can be up to 120ms on a wired LAN connection. After skimming through the code, I found some points that could be improved.

* Screen capturing (on Windows)

The good old GDI approach suffers from performance issues since the introduction of layered windows (Windows 2000). It cannot achieve 60 fps on a 3840x2160 desktop. I found some traces of DXGI desktop duplication in comments, looking forward to seeing the implementation.

* Encoder tuning

There is not so much useful information about the input-to-output delay of libvpx available. I also see some attempts to use NVENC. Generally speaking, it is better to use platform hardware-accelerated encoding API instead. (e.g. Media Foundation, VideoToolbox, VA-API, v4l2m2m)

* Low-latency rendering

This may require playing with platform-native windowing system APIs. (e.g. DXGI_SWAP_EFFECT_FLIP_DISCARD on Windows)

Some game streaming solutions are doing well on this, like Parsec.

**Comment by rustdesk** [comment]
> * Screen capturing

I did using DXGI before, but there was memory leak, so switched to GDI as a workaround.

**Comment by rustdesk** [comment]
I agree with you. Once I get fund support, I will make RustDesk much better.

**Comment by rustdesk** [comment]
https://github.com/H-M-H/Weylus#hardware-acceleration-2

**Comment by rustdesk** [comment]
https://github.com/H-M-H/Weylus/blob/master/lib/encode_video.c

**Comment by rustdesk** [comment]
https://github.com/bryal/dxgcap-rs ref for memory leak

**Comment by rustdesk** [comment]
\display rotation: https://github.com/bryal/dxgcap-rs/blob/master/src/lib.rs#L409

**Comment by rustdesk** [comment]
Split socket read/write

**Comment by AndrewMD5** [comment]
You could also plug in Rainway: https://rainway.com/blog/2022/06/01/free-dev-tier/

**Comment by rustdesk** [comment]
> Rainway

Too many such commercial SDK vendors in the market.

**Comment by AndrewMD5** [comment]
> > Rainway
> 
> Too many such commercial SDK vendors in the market.

Well you mentioned Parsec as inspiration, and Rainway provides that core functionality in a pluggable way while also providing necessary infrastructure. Rainway itself spun out of my old FOSS TeamViewer alternative https://github.com/Ulterius 

So given Rainway is free for FOSS projects and is built on Rust it seemed like a good pairing.

**Comment by rustdesk** [comment]
> Ulterius

Good job. How do you decode video frame on web? We are using vp9 wasm, but it is slow.

**Comment by AndrewMD5** [comment]
> > Ulterius
> 
> Good job. How do you decode video frame on web? We are using vp9 wasm, but it is slow.

VP9 is going to have a lot of overhead for the majority of deployments due the lack of hardware decoding and encoding capabilities on most vendors. If you setup a Rainway account we're happy to allocate time to how you might plug it into your existing solution to take advantage of our real-time media streaming capabilities.
